lib/index.js

```diff
@@ -21,7 +22,59 @@ module.exports.load = function load(rootFileName, options) {
 
     var loadedAbsoluteFileNames = []; // Do some housekeeping to avoid cycles
 
-    return (function readFileAndResolveIncludes(fileName) {
+    // Have to always keep public keys separate within resolvePublic
+    // If a private key is overwritten by a public key, it's an error, and vice-versa. Name your config keys something else!
+    // Empty object leaves and non-public keys are filtered out when requesting with public:true
+
+    function resolvePublic(obj, isToplevel, publicOnly) {
+        if (typeof obj === 'object' && obj !== null) {
+            var publicObj = Array.isArray(obj) ? [] : {};
+
+            Object.keys(obj).forEach(function (key) {
+                if (key !== '#public') {
+                    // Have to extract public and non-public keys separately, otherwise we lose information
+                    var resolvedPublic = resolvePublic(obj[key], false, true);
+                    if (typeof resolvedPublic === 'object' && resolvedPublic !== null && Object.keys(resolvedPublic).length > 0) {
+                        publicObj[key] = resolvedPublic;
+                    }
+
+                    obj[key] = resolvePublic(obj[key], false, false);
+                }
+            });
+
+            if (typeof obj['#public'] !== 'undefined') {
+                if (!(typeof obj['#public'] === 'object' && obj['#public'] !== null)) {
+                    throw new Error('#public must always be an object');
```

Comment:
Arrays are not supported anymore, as the behaviour was not useful. Have added a check and a test for that now.